File: quick-preview-frontend/src/lib/services/translation.ts
/**
 * 번역 서비스 - 배치 처리를 통한 효율적인 번역
 */

import OpenAI from "openai";
import { createLogger } from "@/lib/logger";

const logger = createLogger("Translation");

/** 번역할 세그먼트 */
export interface TranslationSegment {
  start: number;
  end: number;
  text: string;
}

/** 번역된 세그먼트 */
export interface TranslatedSegment {
  start: number;
  end: number;
  originalText: string;
  translatedText: string;
}

/** 배치 번역 설정 */
const BATCH_CONFIG = {
  /** 배치당 세그먼트 수 */
  BATCH_SIZE: 10,
  /** 문맥 유지용 이전 세그먼트 수 */
  CONTEXT_SIZE: 2,
  /** 동시 처리 배치 수 */
  CONCURRENT_BATCHES: 3,
};

/**
 * OpenAI 클라이언트 생성
 */
function getOpenAIClient(): OpenAI {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error("OpenAI API key not configured");
  }
  return new OpenAI({ apiKey });
}

/**
 * 배열을 지정된 크기로 청크 분할
 */
function chunkArray<T>(array: T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < array.length; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}

/**
 * 단일 배치 번역
 */
async function translateBatch(
  openai: OpenAI,
  segments: TranslationSegment[],
  contextText: string = ""
): Promise<string[]> {
  const segmentsJson = segments.map((seg, idx) => ({
    id: idx,
    text: seg.text,
  }));

  const prompt = contextText
    ? `이전 문맥: "${contextText}"\n\n번역할 자막:\n${JSON.stringify(segmentsJson, null, 2)}`
    : `번역할 자막:\n${JSON.stringify(segmentsJson, null, 2)}`;

  const response = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [
      {
        role: "system",
        content: `당신은 영어→한국어 자막 번역 전문가입니다.

규칙:
1. 자연스러운 한국어로 번역하세요
2. 기술 용어는 필요시 원어를 괄호 안에 병기하세요 (예: API(API))
3. 구어체 표현은 자연스럽게 의역하세요
4. 번역 결과만 JSON 배열로 반환하세요

출력 형식:
{
  "translations": [
    {"id": 0, "text": "번역된 텍스트"},
    {"id": 1, "text": "번역된 텍스트"}
  ]
}

JSON만 반환하세요. 다른 설명은 포함하지 마세요.`,
      },
      {
        role: "user",
        content: prompt,
      },
    ],
    temperature: 0.3,
    response_format: { type: "json_object" },
  });

  try {
    const result = JSON.parse(response.choices[0].message.content || "{}");
    const translations = result.translations || [];

    // id 순서대로 정렬하여 텍스트만 추출
    return segments.map((_, idx) => {
      const found = translations.find((t: { id: number; text: string }) => t.id === idx);
      return found?.text || segments[idx].text; // 실패 시 원본 반환
    });
  } catch (error) {
    logger.error("번역 결과 파싱 실패", error);
    return segments.map((seg) => seg.text); // 실패 시 원본 반환
  }
}

/**
 * 전체 세그먼트 배치 번역
 */
export async function translateSegments(
  segments: TranslationSegment[]
): Promise<TranslatedSegment[]> {
  if (segments.length === 0) {
    return [];
  }

  logger.info("번역 시작", { totalSegments: segments.length });

  const openai = getOpenAIClient();
  const batches = chunkArray(segments, BATCH_CONFIG.BATCH_SIZE);
  const translatedSegments: TranslatedSegment[] = [];

  logger.info("배치 생성", {
    batchCount: batches.length,
    batchSize: BATCH_CONFIG.BATCH_SIZE,
  });

  // 배치별 순차 처리 (문맥 유지를 위해)
  let previousContext = "";

  for (let i = 0; i < batches.length; i += BATCH_CONFIG.CONCURRENT_BATCHES) {
    const concurrentBatches = batches.slice(i, i + BATCH_CONFIG.CONCURRENT_BATCHES);

    // 동시에 처리할 배치들
    const batchPromises = concurrentBatches.map(async (batch, batchIdx) => {
      const absoluteBatchIdx = i + batchIdx;

      // 첫 배치가 아니면 이전 문맥 사용
      const context = absoluteBatchIdx === 0 ? "" : previousContext;

      logger.debug(`배치 ${absoluteBatchIdx + 1}/${batches.length} 번역 중`);

      const translations = await translateBatch(openai, batch, context);

      return batch.map((seg, idx) => ({
        start: seg.start,
        end: seg.end,
        originalText: seg.text,
        translatedText: translations[idx],
      }));
    });

    const results = await Promise.all(batchPromises);

    for (const batchResult of results) {
      translatedSegments.push(...batchResult);
    }

    // 다음 배치를 위한 문맥 업데이트 (마지막 2개 세그먼트)
    const lastBatch = concurrentBatches[concurrentBatches.length - 1];
    if (lastBatch && lastBatch.length > 0) {
      const contextSegments = lastBatch.slice(-BATCH_CONFIG.CONTEXT_SIZE);
      previousContext = contextSegments.map((seg) => seg.text).join(" ");
    }

    logger.info(`배치 진행`, {
      completed: Math.min(i + BATCH_CONFIG.CONCURRENT_BATCHES, batches.length),
      total: batches.length,
    });
  }

  logger.info("번역 완료", { translatedCount: translatedSegments.length });

  return translatedSegments;
}

/**
 * 번역이 필요한지 확인
 */
export function needsTranslation(languageCode: string): boolean {
  const koreanCodes = ["ko", "ko-KR"];
  return !koreanCodes.includes(languageCode);
}

File: apps/ai/app/services/translation.py
"""Translation service - Batch translation with OpenAI"""

import json
import logging
import asyncio
from typing import TypedDict
from openai import OpenAI, APIConnectionError, RateLimitError, APIStatusError
from tenacity import (
    retry,
    stop_after_attempt,
    wait_exponential,
    retry_if_exception_type,
    before_sleep_log,
)

from app.config import get_settings
from app.core.exceptions import AIServiceError, ErrorCode

logger = logging.getLogger(__name__)


class SegmentInput(TypedDict):
    start: float
    end: float
    text: str


class TranslatedSegmentOutput(TypedDict):
    start: float
    end: float
    original_text: str
    translated_text: str


# Batch configuration
BATCH_SIZE = 10
CONTEXT_SIZE = 2
CONCURRENT_BATCHES = 3


def get_openai_client() -> OpenAI:
    """Get OpenAI client"""
    settings = get_settings()
    if not settings.openai_api_key:
        raise AIServiceError(
            code=ErrorCode.CONFIGURATION_ERROR,
            message="OpenAI API key not configured",
            status_code=500,
        )
    return OpenAI(api_key=settings.openai_api_key)


def chunk_array(array: list, size: int) -> list[list]:
    """Split array into chunks of specified size"""
    return [array[i:i + size] for i in range(0, len(array), size)]


@retry(
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=1, max=10),
    retry=retry_if_exception_type((APIConnectionError, RateLimitError)),
    before_sleep=before_sleep_log(logger, logging.WARNING),
)
async def translate_batch(
    client: OpenAI,
    segments: list[SegmentInput],
    source_language: str,
    target_language: str,
    context_text: str = "",
) -> list[str]:
    """Translate a single batch of segments"""
    settings = get_settings()

    segments_json = [{"id": idx, "text": seg["text"]} for idx, seg in enumerate(segments)]

    prompt = (
        f'이전 문맥: "{context_text}"\n\n번역할 자막:\n{json.dumps(segments_json, ensure_ascii=False, indent=2)}'
        if context_text
        else f'번역할 자막:\n{json.dumps(segments_json, ensure_ascii=False, indent=2)}'
    )

    # Determine translation direction
    if source_language == "ko" and target_language == "en":
        system_prompt = """당신은 한국어→영어 자막 번역 전문가입니다.

규칙:
1. 자연스러운 영어로 번역하세요
2. 기술 용어는 적절한 영어 용어로 번역하세요
3. 구어체 표현은 자연스럽게 의역하세요
4. 번역 결과만 JSON 배열로 반환하세요

출력 형식:
{
  "translations": [
    {"id": 0, "text": "Translated text"},
    {"id": 1, "text": "Translated text"}
  ]
}

JSON만 반환하세요. 다른 설명은 포함하지 마세요."""
    else:
        system_prompt = """당신은 영어→한국어 자막 번역 전문가입니다.

규칙:
1. 자연스러운 한국어로 번역하세요
2. 기술 용어는 필요시 원어를 괄호 안에 병기하세요 (예: API(API))
3. 구어체 표현은 자연스럽게 의역하세요
4. 번역 결과만 JSON 배열로 반환하세요

출력 형식:
{
  "translations": [
    {"id": 0, "text": "번역된 텍스트"},
    {"id": 1, "text": "번역된 텍스트"}
  ]
}

JSON만 반환하세요. 다른 설명은 포함하지 마세요."""

    try:
        response = client.chat.completions.create(
            model=settings.openai_model,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt},
            ],
            temperature=0.3,
            response_format={"type": "json_object"},
        )

        result = json.loads(response.choices[0].message.content or "{}")
        translations = result.get("translations", [])

        # Sort by id and extract text
        return [
            next(
                (t["text"] for t in translations if t["id"] == idx),
                segments[idx]["text"]  # Fallback to original
            )
            for idx in range(len(segments))
        ]

    except json.JSONDecodeError as e:
        logger.error(f"Failed to parse translation response: {e}")
        return [seg["text"] for seg in segments]
    except (APIConnectionError, RateLimitError):
        raise
    except APIStatusError as e:
        logger.error(f"OpenAI API error: {e}")
        return [seg["text"] for seg in segments]


async def translate_segments(
    segments: list[SegmentInput],
    source_language: str = "en",
    target_language: str = "ko",
) -> list[TranslatedSegmentOutput]:
    """
    Translate all segments with batch processing.

    - Processes segments in batches for efficiency
    - Maintains context across batches
    - Concurrent batch processing
    """
    if not segments:
        return []

    logger.info(f"Starting translation: {len(segments)} segments")

    client = get_openai_client()
    batches = chunk_array(segments, BATCH_SIZE)
    translated_segments: list[TranslatedSegmentOutput] = []

    logger.info(f"Created {len(batches)} batches (size: {BATCH_SIZE})")

    previous_context = ""

    for i in range(0, len(batches), CONCURRENT_BATCHES):
        concurrent_batches = batches[i:i + CONCURRENT_BATCHES]

        # Process batches concurrently
        async def process_batch(batch: list[SegmentInput], batch_idx: int) -> list[TranslatedSegmentOutput]:
            absolute_batch_idx = i + batch_idx
            context = "" if absolute_batch_idx == 0 else previous_context

            logger.debug(f"Translating batch {absolute_batch_idx + 1}/{len(batches)}")

            translations = await translate_batch(
                client, batch, source_language, target_language, context
            )

            return [
                {
                    "start": seg["start"],
                    "end": seg["end"],
                    "original_text": seg["text"],
                    "translated_text": translations[idx],
                }
                for idx, seg in enumerate(batch)
            ]

        tasks = [
            process_batch(batch, batch_idx)
            for batch_idx, batch in enumerate(concurrent_batches)
        ]
        results = await asyncio.gather(*tasks)

        for batch_result in results:
            translated_segments.extend(batch_result)

        # Update context for next batch
        last_batch = concurrent_batches[-1]
        if last_batch:
            context_segments = last_batch[-CONTEXT_SIZE:]
            previous_context = " ".join(seg["text"] for seg in context_segments)

        logger.info(f"Batch progress: {min(i + CONCURRENT_BATCHES, len(batches))}/{len(batches)}")

    logger.info(f"Translation completed: {len(translated_segments)} segments")

    return translated_segments
